Update vertex_renderer properly in PolyEditTool

nitrocalcite opened the pull request:
Proposed (partial) fix for #10670

This adds a listener to the PolyEditTool's first renderer's data source.  The listener updates the `vertex_renderer` appropriately when the data source is updated.

An alternative implementation would add such listeners to _all_ of the renderers present, rather than just the first.  However, I don't think this is required, because:
1) In most (all?) use cases, all the renderers will share the same data source
2) It will always be possible to update the `vertex_renderer` appropriately via the first renderer's data source

This is a partial fix, as it doesn't cover the edge case where you want to change the length of the data source (i.e the number of points) while simultaneously editing with the PolyEditTool.  In this case, #10670 persists in the same manner, and you have to double click the line to fix it.  I think this is a relatively uncommon use case, and it's not made worse due to this PR.  With that being said, perhaps I've misunderstood how the internal logic is supposed to work, giving rise to this behavior.  

Please let me know what you think of this approach and if there are any improvements you'd like me to take a look at.  Thanks!

philippjfr commented on bokehjs/src/lib/models/tools/edit/poly_edit_tool.ts:
```diff
@@ -56,30 +57,48 @@ export class PolyEditToolView extends PolyToolView {
 
+    const vsync_renderer = this.model.renderers[0]
+    const vsync_updater = () => this._update_vertices(vsync_renderer)
+    let vsync_ds = null
+    if (vsync_renderer !== undefined) vsync_ds = vsync_renderer.data_source
```
@mattpap Might correct me here but this might be simpler:

```suggestion
    const vsync_ds = vsync_renderer?.data_source
```

but also necessitates the change below.

mattpap replied:
Definitively.

philippjfr commented on bokehjs/src/lib/models/tools/edit/poly_edit_tool.ts:
```diff
@@ -56,30 +57,48 @@ export class PolyEditToolView extends PolyToolView {
       this._selected_renderer = null
       this._drawing = false
+      this._cur_index = null
+      if (vsync_ds !== null)
```
```suggestion
      if (vsync_ds != null)
```

mattpap replied:
In general one should prefer `!=` over `!==`, unless there is an explicit need for distinguishing between `undefined` and `null`, which should be utilized as rarely as possible.

philippjfr commented on bokehjs/src/lib/models/tools/edit/poly_edit_tool.ts:
```diff
@@ -56,30 +57,48 @@ export class PolyEditToolView extends PolyToolView {
+        vsync_ds.disconnect(vsync_ds.properties.data.change, vsync_updater)
       return
     }
+    if (vsync_ds !== null)
```
```suggestion
    if (vsync_ds != null)
```

philippjfr commented:
First of all thanks! I think this change makes sense, the only thing I might worry about is if a user switches out the data source which would leave a callback on a non-existent data source. That said I think this is sufficiently uncommon.

mattpap commented:
This needs regression tests, otherwise this will be inevitably broken sooner than later.

nitrocalcite commented:
Happy to help, thanks for the quick review!  

I agree that the callback will be left if you change the data source while editing.  I think in general that's a hole - this PR really only helps with updating the data source's values; changing the data source itself or its shape still mildly breaks things a la #10670.

@mattpap I'm happy to provide regression tests but have a dumb question.  I can't figure out how to run just the Python tests for PolyEditTool.  The normal thing:
```
pytest tests/integration/tools/test_poly_edit_tool.py
```
doesn't find any tests, because all of the test functions are named private `_test_something` rather than `test_something`.  I assume that bokeh's testing infrastructure does something with this, but I don't see any code to that effect in `conftest.py`/`setup.cfg`/`bokeh._testing`.   

Apologies if I'm missing something obvious but a pointer would be really helpful.  Thanks!

bryevdv commented:
@nitrocalcite I would say first try to model a new test after the existing tests there, even though they are marked private. IIRC @mattpap mentioned that they mostly seemed to work (but could be improved to avoid the `sleep` calls). I'm not sure how much guidance you need for this so please just let me know if I should elaborate.

nitrocalcite commented:
@bryevdv Exactly my [plan](https://discourse.bokeh.org/t/intentional-suppression-of-all-polyedittool-integration-tests/6787)!  I took some time off for the holidays, and haven't gotten around to it yet.  I'll let you know if I get stuck but don't foresee any problems.  Thanks a lot for checking in & shepherding this though!

nitrocalcite commented:
Changelog:
* Added a regression test, confirmed it indeed fails on `v2.2.3`
* Enabled the PolyEditTool test suite by making the methods public
* Fixed the failing tests & warnings

@bryevdv @mattpap any other tests or work that you'd like to see with this PR?

I'm not certain how defined/rigorous the testing conventions are, and I'm still quite new to bokeh; please feel free to let me know if there are better options than what I've written.  Thanks a lot!

nitrocalcite commented:
@bryevdv if there aren't other changes, any chance this could make it into the 2.3 release?  I've been having to ask people to install my patched and prebuilt fork. 

Thanks!

bryevdv commented:
cc @philippjfr @mattpap

bryevdv commented:
@philippjfr @mattpap it looks like your comments were addressed, any objections to merge this for 2.3?

mattpap commented:
If it works, I'm fine with merging this. Unfortunately I don't have capacity currently to analyze this.

bryevdv commented:
Adding to 2.3 milestone tentatively (would like input from @philippjfr if possible)

philippjfr commented:
Looks good to me. I'll try to test it out before 2.3 is released.

philippjfr commented:
Just tested it, works great! Thanks for your efforts here @nitrocalcite.
